### ISSUE TITLE
The build fails if shared_future<>::then is called with a thread executor

### ISSUE THREAD

**Issue by chinz07** [STEllAR-GROUP/hpx#3634]
## Expected Behavior

The build should succeed.

## Actual Behavior

The build fails with the error
```
/opt/hpx-debug/include/hpx/parallel/executors/thread_execution.hpp:118:18: error: ‘class hpx::threads::executor’ has no member named ‘get_stacksize’
             exec.get_stacksize(),
             ~~~~~^~~~~~~~~~~~~
```
The full error message is:
```
In file included from /opt/hpx-debug/include/hpx/parallel/executors/default_executor.hpp:14,
                 from /opt/hpx-debug/include/hpx/parallel/executors.hpp:11,
                 from /home/christopher/git/work/Qubus/qubus/src/dataflow.cpp:3:
/opt/hpx-debug/include/hpx/parallel/executors/thread_execution.hpp: In instantiation of ‘typename std::enable_if<hpx::traits::is_threads_executor<Policy>::value>::type hpx::threads::post(Executor&&, F&&, Ts&& ...) [with Executor = hpx::threads::executor&; F = std::pair<hpx::threads::thread_state_enum, hpx::threads::thread_id_type> (hpx::lcos::detail::continuation<hpx::lcos::shared_future<void>, hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >, qubus::access_token>::*&)(boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >&&); Ts = {boost::intrusive_ptr<hpx::lcos::detail::continuation<hpx::lcos::shared_future<void>, hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >, qubus::access_token> >, boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >}; typename std::enable_if<hpx::traits::is_threads_executor<Policy>::value>::type = void]’:
/opt/hpx-debug/include/hpx/parallel/executors/execution_fwd.hpp:302:28:   required from ‘decltype (hpx::parallel::execution::detail::post(forward<Executor>(exec), forward<F>(f), (forward<Ts>)(hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::post_tag>::operator()::ts)...)) hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::post_tag>::operator()(Executor&&, F&&, Ts&& ...) const [with Executor = hpx::threads::executor&; F = std::pair<hpx::threads::thread_state_enum, hpx::threads::thread_id_type> (hpx::lcos::detail::continuation<hpx::lcos::shared_future<void>, hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >, qubus::access_token>::*&)(boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >&&); Ts = {boost::intrusive_ptr<hpx::lcos::detail::continuation<hpx::lcos::shared_future<void>, hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >, qubus::access_token> >, boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >}; decltype (hpx::parallel::execution::detail::post(forward<Executor>(exec), forward<F>(f), (forward<Ts>)(hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::post_tag>::operator()::ts)...)) = void]’
/opt/hpx-debug/include/hpx/lcos/local/packaged_continuation.hpp:377:38:   required from ‘void hpx::lcos::detail::continuation<Future, F, ContResult>::async_exec(typename hpx::traits::detail::shared_state_ptr_for<Future>::type&&, Executor&&, hpx::error_code&) [with Executor = hpx::threads::executor&; Future = hpx::lcos::shared_future<void>; F = hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >; ContResult = qubus::access_token; typename hpx::traits::detail::shared_state_ptr_for<Future>::type = boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >]’
/opt/hpx-debug/include/hpx/lcos/local/packaged_continuation.hpp:391:13:   required from ‘void hpx::lcos::detail::continuation<Future, F, ContResult>::async_exec(typename hpx::traits::detail::shared_state_ptr_for<Future>::type&&, Executor&&) [with Executor = hpx::threads::executor&; Future = hpx::lcos::shared_future<void>; F = hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >; ContResult = qubus::access_token; typename hpx::traits::detail::shared_state_ptr_for<Future>::type = boost::intrusive_ptr<hpx::lcos::detail::future_data_base<hpx::traits::detail::future_data_void> >]’
/opt/hpx-debug/include/hpx/lcos/local/packaged_continuation.hpp:485:73:   required from ‘void hpx::lcos::detail::continuation<Future, F, ContResult>::attach_exec(const Future&, typename std::remove_reference<_SrcTuple>::type&) [with Executor = hpx::threads::executor&; Future = hpx::lcos::shared_future<void>; F = hpx::util::detail::one_shot_wrapper<hpx::util::detail::bound_back<qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)> > >; ContResult = qubus::access_token; typename std::remove_reference<_SrcTuple>::type = hpx::threads::executor]’
/opt/hpx-debug/include/hpx/lcos/local/packaged_continuation.hpp:668:9:   [ skipping 2 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/opt/hpx-debug/include/hpx/parallel/executors/execution_fwd.hpp:284:36:   required from ‘decltype (hpx::parallel::execution::detail::then_execute(forward<Executor>(exec), forward<F>(f), forward<Future>(predecessor), (forward<Ts>)(hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::then_execute_tag>::operator()::ts)...)) hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::then_execute_tag>::operator()(Executor&&, F&&, Future&&, Ts&& ...) const [with Executor = hpx::threads::executor&; F = qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)>; Future = hpx::lcos::shared_future<void>; Ts = {}; decltype (hpx::parallel::execution::detail::then_execute(forward<Executor>(exec), forward<F>(f), forward<Future>(predecessor), (forward<Ts>)(hpx::parallel::execution::detail::customization_point<hpx::parallel::execution::detail::then_execute_tag>::operator()::ts)...)) = hpx::lcos::future<qubus::access_token>]’
/opt/hpx-debug/include/hpx/lcos/local/packaged_continuation.hpp:680:49:   required from ‘typename hpx::traits::future_then_executor_result<Executor, typename std::decay<_Arg>::type, F>::type hpx::lcos::detail::then_execute_helper(Executor&&, F&&, Future&&) [with Executor = hpx::threads::executor&; Future = hpx::lcos::shared_future<void>; F = qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)>; typename hpx::traits::future_then_executor_result<Executor, typename std::decay<_Arg>::type, F>::type = hpx::lcos::future<qubus::access_token>]’
/opt/hpx-debug/include/hpx/lcos/future.hpp:477:47:   required from ‘static typename hpx::traits::future_then_executor_result<Executor_, Future, F>::type hpx::lcos::detail::future_then_dispatch<Future, Executor, typename std::enable_if<((hpx::traits::is_one_way_executor<Executor>::value || hpx::traits::is_two_way_executor<Executor>::value) || hpx::traits::is_threads_executor<Executor>::value)>::type>::call(Future&&, Executor_&&, F&&) [with Executor_ = hpx::threads::executor&; F = qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)>; Future = hpx::lcos::shared_future<void>; Executor = hpx::threads::executor; typename hpx::traits::future_then_executor_result<Executor_, Future, F>::type = hpx::lcos::future<qubus::access_token>]’
/opt/hpx-debug/include/hpx/lcos/future.hpp:725:77:   required from ‘static decltype (hpx::lcos::detail::future_base<Derived, R>::future_then_dispatch<typename std::decay<T0>::type>::call(std::move(fut), forward<T0>(t0), forward<F>(f))) hpx::lcos::detail::future_base<Derived, R>::then(Derived&&, T0&&, F&&, hpx::error_code&) [with F = qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)>; T0 = hpx::threads::executor&; Derived = hpx::lcos::shared_future<void>; R = void; decltype (hpx::lcos::detail::future_base<Derived, R>::future_then_dispatch<typename std::decay<T0>::type>::call(std::move(fut), forward<T0>(t0), forward<F>(f))) = hpx::lcos::future<qubus::access_token>; typename std::decay<T0>::type = hpx::threads::executor]’
/opt/hpx-debug/include/hpx/lcos/future.hpp:1385:35:   required from ‘typename hpx::traits::future_then_result<hpx::lcos::shared_future<R>, F>::type hpx::lcos::shared_future<R>::then(hpx::threads::executor&, F&&, hpx::error_code&) const [with F = qubus::dataflow_graph::schedule_modification(const qubus::object&)::<lambda(const hpx::lcos::shared_future<void>&)>; R = void; typename hpx::traits::future_then_result<hpx::lcos::shared_future<R>, F>::type = hpx::lcos::future<qubus::access_token>]’
/home/christopher/git/work/Qubus/qubus/src/dataflow.cpp:99:14:   required from here
/opt/hpx-debug/include/hpx/parallel/executors/thread_execution.hpp:118:18: error: ‘class hpx::threads::executor’ has no member named ‘get_stacksize’
             exec.get_stacksize(),
             ~~~~~^~~~~~~~~~~~~
```

The culprit seems to be the call to the `then` overload at `hpx/lcos/future.hpp:1383` which causes the loss of static type information.

## Steps to Reproduce the Problem

The following code should reproduce the problem:
``` C++
//  Copyright (c) 2019 Christopher Hinz
//
//  Distributed under the Boost Software License, Version 1.0. (See accompanying
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <hpx/hpx_main.hpp>

#include <hpx/lcos/future.hpp>

#include <hpx/parallel/executors.hpp>
#include <hpx/runtime/threads/executors/pool_executor.hpp>

int main()
{
    hpx::threads::executors::pool_executor executor("default");

    auto future = hpx::make_ready_future().share();

    future.then(executor, [](auto future) { future.get(); });

    return 0;
}
```

## Specifications

  - HPX Version: master (15480ea2983f78fc38c792cc0785a5004f210e4f)
  - Platform (compiler, OS): GCC 8.2.1, Arch Linux

**Comment by hkaiser** [comment]
This is most likely a bug. I'll investigate. Thanks!

**Comment by chinz07** [comment]
Hm. Mimicking the structure of `future::then<>` and removing the overload while adding an additional case for thread executors to the generic template does seem to solve the issue. I am still concerned that `hpx::traits::is_threads_executor` only checks if the executor is derived from `executor` but not if it is actually a `scheduled_executor` as required by the implementation.

I will run the entire test suite and report back afterwards.

**Comment by chinz07** [comment]
Except for the usual problems, my changes do not seem to break any tests. I have submitted a PR (#3637) with a possible fix and a test demonstrating the issue.

**Comment by hkaiser** [comment]
This can be closed as #3637 was merged.
